### PR TITLE
Add muon and layerwise distributed optimizer

### PR DESCRIPTION
**Pull request by FDecaYed**
# What does this PR do ?
Add muon and layerwise distributed optimizer to maim

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by JavaZeroo** [comment]
is there any example shell for muon optimizer?

**Comment by BoxiangW** [comment]
@JavaZeroo simply set `--optimizer dist_muon`(Layerwise distributed optimizer) or `--optimizer muon` will turn on Muon usage

**Comment by FDecaYed** [comment]
pullled in https://github.com/NVIDIA/Megatron-LM/pull/2564 and resolved conflicts due to main optimizer refactoring

File: megatron/core/optimizer/muon.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+                )
+                for g in qkv_grads
+            ]
+            grad = torch.cat(qkv_grads, dim=1).view(grad_shape)
+        else:
+            grad = self.scaled_orthogonalize_fn(grad, tp_group, partition_dim)
+        return grad
+
+
+def get_megatron_muon_optimizer(
```
Comment (jstjohn) [inline megatron/core/optimizer/muon.py:165]: Could this be simplified by making use of an idea like https://docs.google.com/document/d/1ZqfINpHhYn9wF3xbIDUvF87tZfS3-RP3Nw1QlTIc1B0/edit?tab=t.0#heading=h.dgnvze2kredp?

You may be able to update the base `get_megatron_optimizer` to use your new `layer_wise_optimizer` kind of like the following:

```
def add_muon_param_groups(config_overrides):
    linear_match_predicate = ParamPredicate("linear_match", fn=lambda param: not getattr(param, 'is_embedding_or_output_parameter', False) and len(param.shape) == 2)
    linear_layer_matcher = ParamKey(predicate=linear_match_predicate)
    config_overrides[linear_layer_matcher] = {"optimizer": "muon"}
    return config_overrides
```
Add those to your config_overrides and then you would have "optimizer": "muon" in the matching param_group dictionaries. So you could look in those and see if you need to do an optimizer override? This would not cover making the custom settings to `param.expert_tp = True` and `param.is_qkv = True` that you have in there now though :/ Maybe those could be added to the param_group, and we could add code in the get_megatron_optimizer function to apply certain keys to all `params` in a `param group` if those magic keys are present in the `param_group` (eg expert_tp, and is_qkv). The problem though is then you'd have a few linear param groups rather than just one. Is that ok?

Then the next thing to do is a for loop over param_group. Something like:

```
optimizers = []
for model_chunk in model_chunks:
    pgs = ...
    for param_group in pgs:
        optimizer_cfg = copy(base_optimizer_cfg)
        for key in param_groups.keys() if key in dir(optimizer_config):
              setattr(optimizer_cfg, key, param_groups[key])
        optimizer = ... # init based on optimizer_cfg.optimizer like we do now, but only with this param_group rather than all
        optimizers.append(optimizer)
final_opt = ... # combine optimizers into one
```

**Comment by ko3n1g** [comment]
/ok to test [cd3fe36](https://github.com/NVIDIA/Megatron-LM/pull/2241/commits/cd3fe364549ba19a365f98fffab76257c2854670)

File: megatron/core/optimizer/muon.py (hunk 1 above)
Comment (BoxiangW) [reply megatron/core/optimizer/muon.py:165]: Yes we can move towards this design next release

Comment (deepakn94) [reply megatron/core/optimizer/muon.py:165]: Who is going to take the action item to move towards this design next release? Should we create an issue?

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -1217,6 +1217,13 @@ def validate_args(args, defaults={}):
         if args.rank == 0:
             print('Warning: enabling --no-load-optim when skipping training.')
 
+    # Muon optimizercheck
+    if 'muon' in args.optimizer:
```
Comment (deepakn94) [inline megatron/training/arguments.py:1221]: How about making sure `overlap_grad_reduce` and `overlap_param_gather` are False?

Comment (BoxiangW) [reply megatron/training/arguments.py:1221]: Added assertion for them

File: megatron/core/optimizer/muon.py (hunk 1 above)
Comment (BoxiangW) [reply megatron/core/optimizer/muon.py:165]: I have created a issue here for tracking https://github.com/NVIDIA/Megatron-LM/issues/2936

File: megatron/core/optimizer/optimizer_config.py [hunk 3]
```diff
@@ -198,6 +198,31 @@ class OptimizerConfig:
     sgd_momentum: float = 0.9
     """Momentum factor for SGD optimizer."""
 
+    # Muon
+    muon_momentum: float = 0.95
```
Comment (deepakn94) [inline megatron/core/optimizer/optimizer_config.py:203]: Add TODO to create a separate MuonConfig?

**Review by deepakn94 — APPROVED** [review]

File: megatron/core/optimizer/optimizer_config.py (hunk 3 above)
Comment (Phlip79) [reply megatron/core/optimizer/optimizer_config.py:203]: Added TODO.

Comment (BoxiangW) [reply megatron/core/optimizer/optimizer_config.py:203]: Added note in the issue as well
